File: mytradebot/memory/read/citations.py
"""文件用途：解析 Codex 风格的 memory citation 块。"""
from __future__ import annotations

from dataclasses import dataclass
import re


_CITATION_BLOCK_RE = re.compile(r"<oai-mem-citation>\s*(.*?)\s*</oai-mem-citation>", re.DOTALL)
_SECTION_RE = re.compile(r"<(?P<name>citation_entries|rollout_ids)>\s*(?P<body>.*?)\s*</(?P=name)>", re.DOTALL)
_ENTRY_RE = re.compile(
    r"^(?P<path>[^:\n]+):(?P<start>\d+)-(?P<end>\d+)\|note=\[(?P<note>[^\]\n]*)\]$"
)


@dataclass(frozen=True)
class MemoryCitationEntry:
    """说明：一条 memory 文件引用。"""

    file_path: str
    line_start: int
    line_end: int
    note: str


@dataclass(frozen=True)
class MemoryCitations:
    """说明：一个 `<oai-mem-citation>` 块的解析结果。"""

    entries: tuple[MemoryCitationEntry, ...]
    rollout_ids: tuple[str, ...]


def parse_memory_citations(content: str) -> MemoryCitations | None:
    """说明：解析回答末尾的 Codex 风格 memory citation 块。"""
    matched = _CITATION_BLOCK_RE.search(content or "")
    if not matched:
        return None
    sections = {
        section.group("name"): section.group("body")
        for section in _SECTION_RE.finditer(matched.group(1))
    }
    entries = tuple(
        entry
        for line in sections.get("citation_entries", "").splitlines()
        if (entry := _parse_entry(line.strip())) is not None
    )
    rollout_ids = tuple(
        dict.fromkeys(
            line.strip()
            for line in sections.get("rollout_ids", "").splitlines()
            if _looks_like_uuid(line.strip())
        )
    )
    return MemoryCitations(entries=entries, rollout_ids=rollout_ids)


def _parse_entry(line: str) -> MemoryCitationEntry | None:
    matched = _ENTRY_RE.match(line)
    if not matched:
        return None
    line_start = int(matched.group("start"))
    line_end = int(matched.group("end"))
    if line_end < line_start:
        return None
    return MemoryCitationEntry(
        file_path=matched.group("path").strip(),
        line_start=line_start,
        line_end=line_end,
        note=matched.group("note").strip(),
    )


def _looks_like_uuid(value: str) -> bool:
    return bool(re.fullmatch(
        r"[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}",
        value,
    ))


__all__ = [
    "MemoryCitationEntry",
    "MemoryCitations",
    "parse_memory_citations",
]

File: tradex/memory/read/citations.ts
const CITATION_BLOCK_RE = /<oai-mem-citation>\s*([\s\S]*?)\s*<\/oai-mem-citation>/;
const SECTION_RE = /<(?<name>citation_entries|rollout_ids)>\s*(?<body>[\s\S]*?)\s*<\/\k<name>>/g;
const ENTRY_RE = /^(?<path>[^:\n]+):(?<start>\d+)-(?<end>\d+)\|note=\[(?<note>[^\]\n]*)\]$/;
const UUID_RE = /^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/;

export interface MemoryCitationEntry {
  filePath: string;
  lineStart: number;
  lineEnd: number;
  note: string;
}

export interface MemoryCitations {
  entries: MemoryCitationEntry[];
  rolloutIds: string[];
}

export function parseMemoryCitations(content: string): MemoryCitations | null {
  const blockMatch = CITATION_BLOCK_RE.exec(content || "");
  if (!blockMatch) return null;

  const sections: Record<string, string> = {};
  let sectionMatch: RegExpExecArray | null;
  const sectionRe = new RegExp(SECTION_RE.source, "g");
  while ((sectionMatch = sectionRe.exec(blockMatch[1])) !== null) {
    if (sectionMatch.groups) {
      sections[sectionMatch.groups.name] = sectionMatch.groups.body;
    }
  }

  const entries: MemoryCitationEntry[] = [];
  for (const line of (sections.citation_entries ?? "").split("\n")) {
    const parsed = parseEntry(line.trim());
    if (parsed) entries.push(parsed);
  }

  const seen = new Set<string>();
  const rolloutIds: string[] = [];
  for (const line of (sections.rollout_ids ?? "").split("\n")) {
    const trimmed = line.trim();
    if (UUID_RE.test(trimmed) && !seen.has(trimmed)) {
      seen.add(trimmed);
      rolloutIds.push(trimmed);
    }
  }

  if (entries.length === 0 && rolloutIds.length === 0) return null;
  return { entries, rolloutIds };
}

function parseEntry(line: string): MemoryCitationEntry | null {
  if (!line) return null;
  const match = ENTRY_RE.exec(line);
  if (!match?.groups) return null;
  const lineStart = parseInt(match.groups.start, 10);
  const lineEnd = parseInt(match.groups.end, 10);
  if (lineEnd < lineStart) return null;
  return {
    filePath: match.groups.path.trim(),
    lineStart,
    lineEnd,
    note: match.groups.note.trim(),
  };
}

export function formatCitation(citation: MemoryCitationEntry): string {
  return `${citation.filePath}:${citation.lineStart}-${citation.lineEnd}|note=[${citation.note}]`;
}
